docs: Light Component - edit pass (#764)
BuilderIO/qwik: packages/docs/src/pages/tutorial/component/inline/index.mdx

layout: tutorial
---

> Qwik components are lazy-loadable, which lies at the heart of what makes Qwik different from most other frameworks.

Reviewer: Yes totally agree. 

Qwik components are:
- Lazy-loadable=> But all frameworks claim that => Qwik can lazy-load even the components that are in the current DOM (other frameworks can only lazy-load components that are currently not in the DOM)
- Can be lazy-loaded out of order => Meaning a middle component can be resumed without waking up the parent or a child component. (Other frameworks always lazy from root towards children, and they can never skip over things)
- Loaded on interaction (Other frameworks always have to eagerly load code before interaction)
- But it it is not just components. Qwik really lazy loads any closure in this way. Event listeners, effects, behavior etc...

> Qwik components are lazy-loadable, which lies at the heart of what makes Qwik different from most other frameworks.

Reviewer 2: Thanks @mhevery !

@craigshoemaker what are your thoughts? you're the docs master here :)
Can you think of a good way to incoporate this feedback in this doc? 

thanks 🙏

> Qwik components are lazy-loadable, which lies at the heart of what makes Qwik different from most other frameworks.

Author: ```suggestion
One of Qwik's super powers lies in its lazy-loading features. 

Qwik lazy-loads:

- **Independent of DOM presence**: Components can load before and after they are added to the DOM.
- **Independent of hierarchy**: Components can be loaded out of order. For example, the script for a child component can load before its parent script.
- **Based on interaction**: Script loading is deferred until a user interacts with a component.
- **More than just components**: Qwik lazy-loads any closure including components, event listeners, effects, and behaviors.
```

> Qwik components are lazy-loadable, which lies at the heart of what makes Qwik different from most other frameworks.

Reviewer 2: The new changes based on this thread LGTM
Thanks @craigshoemaker !

Let's see what Misko is thinking on the other thread, I think we should add the rules, WDYT Craig?

> Qwik components are lazy-loadable, which lies at the heart of what makes Qwik different from most other frameworks.

Author: @shairez - they are now in the text 🎉
@mhevery - feel free to merge if you like the current version

> You can mark a component as lazy-loadable by using the `component$()` method. When you see a `$` in Qwik code, you're encountering a lazy-loading boundary.

Reviewer 2: @craigshoemaker I would change the word "encountering" to "entering" like so -  "when you see a '$' in Qwik code, you are entering a lazy loaded boundry. "

@mhevery WDYT?

Reviewer 2: @mhevery thanks!
Maybe it's a good point to mention the limitations of what "crossing the boundary" means? (to give more weight to it, so developers will know "this is an important point, I should make a mental note here")

WDYT?

Reviewer: Yes! Every time you see `$`, you (as a developer) should be thinking special rules apply here. Not every valid JS is valid Qwik code.

Reviewer 2: @mhevery do you think we should add these rules in this section as well?

Reviewer: Hmm, they should probably be both here and in the https://qwik.builder.io/docs/advanced/optimizer!
```
- all captured variables must be declared `const`
- all captured variables must be either:
  - serializable
  - importable (either `import` or `export` in this file)
```

Reviewer 2: Awesome!
Thanks @mhevery

Author: Here's #777 to add these rules to the Optimizer doc 😎

layout: tutorial
---

One of Qwik's super powers lies in its lazy-loading features. 

Qwik lazy-loads:

> - **Independent of DOM presence**: Components can load before and after they are added to the DOM.

Reviewer: ```suggestion
```
Not sure what you were going for, but the above does not make sense without the context of Qwik. So I would just remove it.

Author: It was an attempt to work with this comment:

> Lazy-loadable=> But all frameworks claim that => Qwik can lazy-load even the components that are in the current DOM (other frameworks can only lazy-load components that are currently not in the DOM)

perhaps there's another way to word it?

> - **Independent of hierarchy**: Components can be loaded out of order. For example, the script for a child component can load before its parent script.

Reviewer: ```suggestion
- **Independent of hierarchy**: Components can be loaded out of order. For example, the code for a child component can load before its parent code.
```

> - **Based on interaction**: Script loading is deferred until a user interacts with a component.

Reviewer: ```suggestion
- **Based on interaction**: Code loading is deferred until a user interacts with a component.
```

- **More than just components**: Qwik lazy-loads any closure including components, event listeners, effects, and behaviors.

> You can mark a component as lazy-loadable by using the `component$()` method. When you see a `$` in Qwik code, you're crossing a lazy-loading boundary.

Reviewer: ```suggestion
`$` marks a closure as lazy-loadable. For example `component$()` method makes the component lazy-loadable. When you see a `$` in Qwik code, you're crossing a lazy-loading boundary and have to be aware of special rules:
- any lexically scoped variables must be declared as `const`
- a captured variable/symbol must be either:
  - serializable 
  - importable (Either from a different file as `import` or from this file using `export`)
```
